src/main/java/io/temporal/common/converter/EncodedValues.java

```diff
@@ -58,22 +58,35 @@ public void setDataConverter(DataConverter converter) {
   }
 
   @Override
-  public <T> T get(Class<T> parameterType) throws DataConverterException {
-    if (value != null) {
-      @SuppressWarnings("unchecked")
-      T result = (T) value.orElse(null);
-      return result;
+  public int getSize() {
+    if (values != null) {
+      return values.length;
     } else {
-      if (converter == null) {
-        throw new IllegalStateException("converter not set");
+      if (payloads.isPresent()) {
+        return payloads.get().getPayloadsCount();
+      } else {
+        return 0;
       }
-      return converter.fromPayloads(payloads, parameterType, parameterType);
     }
   }
 
   @Override
-  public <T> T get(Class<T> parameterType, Type genericParameterType)
+  public <T> T get(int index, Class<T> parameterType) throws DataConverterException {
+    return get(index, parameterType, parameterType);
+  }
+
+  @Override
+  public <T> T get(int index, Class<T> parameterType, Type genericParameterType)
       throws DataConverterException {
-    return converter.fromPayloads(payloads, parameterType, genericParameterType);
+    if (values != null) {
+      @SuppressWarnings("unchecked")
+      T result = (T) values[index];
+      return result;
+    } else {
+      if (converter == null) {
+        throw new IllegalStateException("converter not set");
+      }
+      return converter.fromPayloads(index, payloads, parameterType, genericParameterType);
```

Comment:
The idea of Values is that it can be created with objects or Payloads. I'm not sure if caching is really going to help here as every value accessed only once I believe.